### PR TITLE
Update LanguageHelper.php to handle translated text with embedded new lines.

### PR DESCRIPTION
**Pull request by BrainforgeUK**
Emulate Joomla v4.4.0 and earlier behaviour when multi-line text strings present.

Pull Request for Issue #42416 .

### Summary of Changes

Allow embedded newlines in text (i.e. between quotes ").
Using them to structure very large areas of text makes maintenance easier.
In some cases using linewrap in IDE slightly helps, but one is left with trying to read / unravel a massive jumble of text.

Newlines in text worked up to v4.4.0
Security related change in v4.4.1 removed this functionality.

This change detects and removes the embedded newlines before using PHP ini parsing as before.
The security related change introduced in v4.4.1 is maintained.

### Testing Instructions

Install this plugin.
[plg_system_bflanguagetest.zip](https://github.com/joomla/joomla-cms/files/13539524/plg_system_bflanguagetest.zip)

Go to plugin admin page.

### Actual result BEFORE applying this Pull Request

The plugin name and description not translated.
The language labels are displayed.

### Expected result AFTER applying this Pull Request

Name and description translated.
As previous behavious in J4.1.0 and earlier.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

Line breaks in characters string have worked for a very long time.
Mention made in another conversation that somewhere in the documentation it says they are not allowed.
Is so the location where that appears needs to be found and corrected.

Also mention was made that the JED needs to be changed to prevent the use of such line breaks.
Such a change will no longer be necessary - assume the developer wants such included.

### CODE REVIEW

File: libraries/src/Language/LanguageHelper.php [hunk 1]
```diff
@@ -404,33 +404,43 @@ public static function parseIniFile($fileName, $debug = false)
...
-        if (!\function_exists('parse_ini_file') || $isParseIniFileDisabled) {
-            $contents = file_get_contents($fileName);
-            $strings  = @parse_ini_string($contents, false, INI_SCANNER_RAW);
-        } else {
-            $strings = @parse_ini_file($fileName, false, INI_SCANNER_RAW);
-        }
+        // Capture hidden PHP errors from the parsing.
+        set_error_handler(static function ($errno, $err) {
+            throw new \Exception($err);
+        }, \E_WARNING);
```
Comment (brianteeman) [inline libraries/src/Language/LanguageHelper.php:415]: What is the reasoning behind this change. If I read it correctly the original code only captured the errors when debug was on. Now it is always capturing.

File: libraries/src/Language/LanguageHelper.php [hunk 2]
```diff
@@ -404,33 +404,43 @@ public static function parseIniFile($fileName, $debug = false)
...
+                $contents = file_get_contents($fileName);
+                $strings  = parse_ini_string($contents, false, INI_SCANNER_RAW);
+            } else {
+                $strings = parse_ini_file($fileName, false, INI_SCANNER_RAW);
+            }
+        } catch (\Exception $e) {
+            // Emulate Joomla 4.4.0 and earlier handling of multi-line text strings.
+            // Some developers prefer to format long descriptive strings in that way for easier maintenance.
+            // For performance reasons we only do this as a last resort as most .ini files are not affected by this.
+            $strings = self::parseMultilineIni(isset($contents) ? $contents : file_get_contents($fileName));
```
Comment (laoneo) [inline libraries/src/Language/LanguageHelper.php:428]: Did you forgot to add this function?

Comment (BrainforgeUK) [reply libraries/src/Language/LanguageHelper.php:428]: Bizarre ... fixed!

**Comment by dryabov** [comment]
I thought since you're trying to fix it, you'd understand the exact reason for @bembelimen's patch (https://github.com/joomla/joomla-cms/commit/609cafcb752911d264608cf520ad2bf215c72e75), and the problem is that I don't, and it looks like you're trying to fix the problem that shouldn't exist.

**Comment by BrainforgeUK** [comment]
The problem did not exist until J4.1.1

The security fix in J4.1.1 meant that some previously working language .ini files no longer worked.

Without arguing about whether the security issue is real or imagined (in the language translation scenario) and reversing the change the quickest solution is to cater for those failing .ini files without impacting performance with the others... unless you argue that finding all the problematic .ini files (plus months of more come to light as people upgrade to J4.1.1+) is the correct solution - but then those problem .ini files I've encountered have been working for years without a problem.

**Comment by brianteeman** [comment]
I will leave it to the JSST to respond specifically about the _security issue_ but the explanation of why to use parse_ini_raw is well documented

File: libraries/src/Language/LanguageHelper.php (hunk 1 above)
Comment (BrainforgeUK) [reply libraries/src/Language/LanguageHelper.php:415]: I took the exception / debug handling from J5.0.0 ... best to use that
Embedded new lines cause an exception.
Before checking for debug call my custom parsing,
If successful OK, or check for debug mode and handle as before.

**Comment by BrainforgeUK** [comment]
No point waiting for JSST to respond, I am happy with what has been done for J5.0
My changes are based upon J5.0 as it is cleaner error handling.

**Comment by MacJoom** [comment]
@BrainforgeUK thanks for the PR - could you fix the phpcs error?

File: libraries/src/Language/LanguageHelper.php [hunk 3]
```diff
@@ -404,38 +404,129 @@ public static function parseIniFile($fileName, $debug = false)
...
+                {
+                    throw new \RuntimeException($e->getMessage());
+                }
+
+                return [];
+            }
 
-        // Ini files are processed in the "RAW" mode of parse_ini_string, leaving escaped quotes untouched - lets postprocess them
-        $strings = str_replace('\"', '"', $strings);
+            // Ini file processing has left escaped quotes untouched - lets postprocess them
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
            // Ini file processing has left escaped quotes untouched - let's postprocess them
```

File: libraries/src/Language/LanguageHelper.php [hunk 4]
```diff
@@ -404,38 +404,129 @@ public static function parseIniFile($fileName, $debug = false)
...
+     * Parse strings from a language file.
+     * See issue https://github.com/joomla/joomla-cms/issues/42416
+     * Emulates Joomla v4.4.0 and earlier behaviour with multi-line text strings without the potential security concerns.
+     *
+     * @param   string   $fileName  The language ini file path.
+     * @param   boolean  $debug     If set to true debug language ini file.
+     *
+     * @return  array | false  The strings parsed.
+     *
+     * @since   4.1.1
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]: ??? or `__DEPLOY_VERSION__`

File: libraries/src/Language/LanguageHelper.php (hunk 3 above)
Comment (BrainforgeUK) [reply libraries/src/Language/LanguageHelper.php]: Committed change.

File: libraries/src/Language/LanguageHelper.php (hunk 4 above)
Comment (BrainforgeUK) [reply libraries/src/Language/LanguageHelper.php]: Now uses __DEPLOY_VERSION__

**Comment by BrainforgeUK** [comment]
Anything else I need to do here?
**This branch is out-of-date with the base branch**
Do I have to do the _Update branch_ with 4.4-dev? Joomla 5?

**Comment by MacJoom** [comment]
> Anything else I need to do here? **This branch is out-of-date with the base branch** Do I have to do the _Update branch_ with 4.4-dev? Joomla 5?

No - i did the Update branch (comes up if the current branch - 4.4- is updated with other PR's) - it looks ok now

**Comment by BrainforgeUK** [comment]
Thanks.
However, I cannot see what the phpcs error is referring to.
https://ci.joomla.org/joomla/joomla-cms/72531/1/8

File: libraries/src/Language/LanguageHelper.php [hunk 5]
```diff
@@ -404,38 +404,127 @@ public static function parseIniFile($fileName, $debug = false)
...
+        $lines = explode("\n", $inifileContents);
+
+        $nameValuePairs = [];
+
+        for ($l=0; $l<count($lines); $l++) {
+            $line = trim($lines[$l]);
+
+            if (empty($line)) continue;
+
+            if ($line[0] == ';') continue;
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
            if (empty($line)) {
                continue;
            }

            if ($line[0] == ';') {
                continue;
            }
```

File: libraries/src/Language/LanguageHelper.php [hunk 6]
```diff
@@ -404,38 +404,127 @@ public static function parseIniFile($fileName, $debug = false)
...
+     * @return  array | false  The strings parsed.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected static function parseMultilineIni($inifileContents) {
+        $lines = explode("\n", $inifileContents);
+
+        $nameValuePairs = [];
+
+        for ($l=0; $l<count($lines); $l++) {
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
        for ($l=0; $l < count($lines); $l++) {
```

File: libraries/src/Language/LanguageHelper.php [hunk 7]
```diff
@@ -404,38 +404,127 @@ public static function parseIniFile($fileName, $debug = false)
...
+            }
+
+            if (str_ends_with($lineParts[1], '"')) {
+                // Just in case we have a line ending with an escaped "
+                if (!str_ends_with($line, '\\"')) {
+                    if (strlen($lineParts[1]) == 2) {
+                        $nameValuePairs[$lineParts[0]] = '';
+                        continue;
+                    }
+                    $nameValuePairs[$lineParts[0]] = substr($lineParts[1], 1, strlen($lineParts[1])-2);
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
                    $nameValuePairs[$lineParts[0]] = substr($lineParts[1], 1, strlen($lineParts[1]) - 2);
```

File: libraries/src/Language/LanguageHelper.php [hunk 8]
```diff
@@ -404,38 +404,127 @@ public static function parseIniFile($fileName, $debug = false)
...
+
+            while (++$l<count($lines)) {
+                $line = trim($lines[$l]);
+
+                $nameValuePair .= "\n" . $line;
+
+                if (str_ends_with($line, '"')) {
+
+                    if (!str_ends_with($line, '\\"')) {
+                        $nameValuePairs[$lineParts[0]] = substr($nameValuePair, 1, strlen($nameValuePair)-1);
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
            while (++$l < count($lines)) {
                $line = trim($lines[$l]);

                $nameValuePair .= "\n" . $line;

                if (str_ends_with($line, '"')) {
                    if (!str_ends_with($line, '\\"')) {
                        $nameValuePairs[$lineParts[0]] = substr($nameValuePair, 1, strlen($nameValuePair) - 1);
```

File: libraries/src/Language/LanguageHelper.php [hunk 9]
```diff
@@ -404,38 +404,127 @@ public static function parseIniFile($fileName, $debug = false)
...
+     * Emulates Joomla v4.4.0 and earlier behaviour with multi-line text strings without the potential security concerns.
+     *
+     * @param   string   $fileName  The language ini file path.
+     * @param   boolean  $debug     If set to true debug language ini file.
+     *
+     * @return  array | false  The strings parsed.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected static function parseMultilineIni($inifileContents) {
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
    protected static function parseMultilineIni($inifileContents)
    {
```

File: libraries/src/Language/LanguageHelper.php [hunk 10]
```diff
@@ -404,38 +404,130 @@ public static function parseIniFile($fileName, $debug = false)
...
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected static function parseMultilineIni($inifileContents)
+    {
+        $lines = explode("\n", $inifileContents);
+
+        $nameValuePairs = [];
+
+        for ($l=0; $l < count($lines); $l++) {
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]:
```suggestion
        for ($l = 0; $l < count($lines); $l++) {
```

Comment (BrainforgeUK) [reply libraries/src/Language/LanguageHelper.php]: Thanks, I see now how to interpret phpcs errors.
That last one was the clearest yet.

**Comment by bembelimen** [comment]
Hello @BrainforgeUK we, the maintainer team decided, that the solution is not the way we want to go.

Thank you very much for your contribution, we really appreciate your work and hope, this does not discourage you and we're looking forward to further ideas from you.

**Comment by BrainforgeUK** [comment]
Thanks, my concern is when will embedded newline capability be restored to .ini language files?
My solution would have at least provided a temporary resolution until a more acceptable longer term solution (and I assume pre v4.4.1 backward compatible)) implemented.
Starting work today on a new component and encountered how inconvenient this is going to be.
Wordwrap in the IDE partly helps but is not a long-term solution.

**Pull request closed without merging**